bundles/alpha.model/src/alpha/model/util/Facet.xtend

```diff
@@ -77,18 +86,82 @@ class Facet {
+		equalities = DomainOperations.toISLEqualityMatrix(basicSetNoRedundancies)
+		indexCount = basicSetNoRedundancies.dim(ISLDimType.isl_dim_set)
+		
+		val allInequalities = getInequalities(basicSetNoRedundancies, indexCount, true)
+		val inequalities = separateEffectivelySaturatedInequalities(allInequalities, space)
+		indexInequalities = inequalities.key
+		effectivelySaturatedInequalities = inequalities.value
+		
+		if (effectivelySaturatedInequalities.nbRows % 2 != 0) {
+			throw new Exception('Failed to create lattice, there should be an even number of (or zero) effectively saturated constraints')
+		}
+		
 		indexInequalityCount = indexInequalities.nbRows
-		isBounded = basicSet.bounded
-		isEmpty = basicSet.empty
+		isBounded = basicSetNoRedundancies.bounded
+		isEmpty = basicSetNoRedundancies.empty
 		parameterEqualityCount = countParameterConstraints(equalities, indexCount)
-		parameterInequalities = getInequalities(basicSet, indexCount, false)
+		parameterInequalities = getInequalities(basicSetNoRedundancies, indexCount, false)
 		this.saturatedInequalityIndices = saturatedInequalityIndices
-		space = basicSet.space 
+		
+		dimensionality = dimensionality(effectivelySaturatedInequalities, equalities, indexCount)
+	}
+	
+	/** Separates the effectively saturated inequalities from true unsaturated inequalities */
+	def static separateEffectivelySaturatedInequalities(ISLMatrix allInequalities, ISLSpace space) {
+		val matrix = allInequalities.copy.dropCols(space.nbParams + space.nbIndices, 1).toLongMatrix
+
+		// Use a hash map to get the index of a particular vector in matrix
+		val idxMap = new HashMap<String, Long>()
+		(0..<matrix.length).forEach[i | idxMap.put(matrix.get(i).toString, Long.valueOf(i))]
+
+		
+		// Identify the indices of any parallel constraints. For each vector in matrix, check
+		// to see if its negation exists (i.e., is a key in idxMap). Label such vectors as 
+		// having a colinear partner.
+		val negMatrix = matrix.length > 0 ? matrix.scalarMultiplication(-1) : matrix
+		
+		val colinearConstraints = new HashSet<Integer>()
 
-		dimensionality = dimensionality(equalities, indexCount)
+		(0..<matrix.length).forEach[i |
+			val constraintVector = negMatrix.get(i).toString
+			val hasConlinearPartner = idxMap.containsKey(constraintVector)
+			if (hasConlinearPartner) {
+				colinearConstraints += i
+			}
+		]
+		
+		// Construct the separate matrix for all colinear inequalities
+		val longMatrix = allInequalities.toLongMatrix
+		val effectivelySaturatedInequalities = (0..<matrix.length)
+			.filter[i | colinearConstraints.contains(i)]
+			.map[i | longMatrix.get(i)]
+			.buildFromLongMatrixEvenIfEmpty(space)
+		
+		// Drop all conlinear inequalities from indexInequalities
+//		val unsaturatedInequalities = colinearConstraints.fold(allInequalities, [
+//			mat, row | mat.dropRows(row, 1)
+//		])
+		val unsaturatedInequalities = (0..<matrix.length)
+			.filter[i | ! colinearConstraints.contains(i)]
```

Comment:
For the `unsaturatedInequalities` computation inside `separateEffectivelySaturatedInequalities`, instead of calling `filter` with a negated predicate, you can call `reject` and avoid the negation as such:

```
(0 ..< matrix.length)
    .reject[i | colinearConstraints.contains(i)]
    ...
`